packages/world/src/utils/InventoryUtils.sol

```diff
@@ -100,9 +100,37 @@ library InventoryUtils {
   function addEntityToSlot(EntityId owner, EntityId entityId, uint16 slot) internal {
     uint16 maxSlots = EntityObjectType._get(owner).getMaxInventorySlots();
     require(slot < maxSlots, "Invalid slot");
-    require(entityId.exists(), "Entity must exist");
+    ObjectType objectType = EntityObjectType._get(entityId);
+    require(!objectType.isNull(), "Entity must exist");
+
+    InventorySlotData memory slotData = InventorySlot._get(owner, slot);
+
+    require(slotData.objectType.isNull(), "Slot must be empty");
+
+    // Entities always have amount 1
+    InventorySlot._setAmount(owner, slot, 1);
```

Comment:
What was happening before when we didnt do all these writes?

Reply:
addEntityToSlot is only used in transfer, and it would work for most cases, except for when you do:

- add object to slot
- remove it
- transfer a tool from a differnt slot to this slot
- in this case the occupied slot would not be set